### PR TITLE
Fix more warnings, and possible resource leaks

### PR DESCRIPTION
**Pull request by drwill-ms**
Updated to latest FxCop nuget. That fixed some false warnings, but introduces new standards.

Two such standards that we consistently break is not having hard-coded strings as error messages, and catching System.Exception. Rather than fix all those (ha) or suppress each one individually, I chose to add globalsuppression files to each project.

Most warning fixes are:

- using nameof for parameter validation exception params
- parameter null checking
- calling dispose on IDisposable objects just before they would be out of scope (oops! - pay special attention to these changes please)
- adding some missed parameter xml tags
- methods marked as async that didn't make await anything
- some exceptions had a public ctor with serialization info param, which apparently shouldn't be public. As we can't change the public API, these have been suppressed.
- compare strings using StringComparer.Ordinal/Invariant.Equals
- using string.IsNullOrWhiteSpace or equiv
- do not use Count() method if Length or Count properties are available

Also included are style changes including:

- usings outside of namespace
- using auto properties where possible
- auto-property getter and setter all on one line
- simply property getter or setter use lambda notation for fewer lines
- class member names, visibility, and readonly when appropriate
- removing "this." now that ambiguity is gone with field names
- formatting method calls and parameters for easier reading, when they get long

### CODE REVIEW

File: provisioning/transport/amqp/src/ProvisioningTransportHandlerAmqp.cs [hunk 1]
```diff
@@ -191,66 +196,88 @@ private static async Task CreateLinksAsync(AmqpClientConnection connection, stri
             string correlationId,
             DeviceRegistration deviceRegistration)
         {
-            AmqpMessage amqpMessage;
-            if (deviceRegistration == null)
+            AmqpMessage amqpMessage = null;
+
+            try
             {
-                amqpMessage = AmqpMessage.Create(new MemoryStream(), true);
```
Comment (drwill-ms) [inline provisioning/transport/amqp/src/ProvisioningTransportHandlerAmqp.cs]: lots of resource leaks in this file :(

File: provisioning/transport/mqtt/src/ProvisioningChannelHandlerAdapter.cs [hunk 2]
```diff
@@ -293,12 +299,13 @@ private async Task PublishRegisterAsync(IChannelHandlerContext context)
             if (_message.Payload != null && _message.Payload.Length > 0)
             {
                 var deviceRegistration = new DeviceRegistration { Payload = new JRaw(_message.Payload) };
-                var customContentStream =
```
Comment (drwill-ms) [inline provisioning/transport/mqtt/src/ProvisioningChannelHandlerAdapter.cs]: Another resource leak

File: e2e/test/FaultInjectionPoolAmqpTests.MessageSendFaultInjectionPoolAmqpTests.cs [hunk 3]
```diff
@@ -821,27 +821,31 @@ public async Task Message_IoTHubSak_AuthenticationNoRecovery_MultipleConnections
...
-                devicesCount,
-                faultType,
-                reason,
-                delayInSec,
-                durationInSec,
-                (d, t) => { return Task.FromResult<bool>(false); },
-                testOperation,
-                cleanupOperation,
-                authScope).ConfigureAwait(false);
+            await FaultInjectionPoolingOverAmqp
```
Comment (vinagesh) [inline e2e/test/FaultInjectionPoolAmqpTests.MessageSendFaultInjectionPoolAmqpTests.cs]: minor: I don't know if this is public facing or not but async function name does not seem to be ending with async

File: iothub/device/src/Common/Exceptions/IotHubException.cs [hunk 4]
```diff
@@ -18,41 +18,15 @@ public class IotHubException : Exception
...
         /// <summary>
         /// Gets a value indicating if the error is transient.
         /// </summary>
-        public bool IsTransient
-        {
-            get
-            {
-                return _isTransient;
-            }
-            private set
```
Comment (barustum) [inline iothub/device/src/Common/Exceptions/IotHubException.cs]: changing from private set to set. Intentional?

File: iothub/device/src/Common/Security/SharedAccessSignature.cs [hunk 5]
```diff
@@ -140,29 +103,29 @@ public static bool IsSharedAccessSignature(string rawSignature)
...
+            return ExpiresOn + SharedAccessSignatureConstants.MaxClockSkew < DateTime.UtcNow;
         }
 
         public void Authenticate(SharedAccessSignatureAuthorizationRule sasAuthorizationRule)
         {
-            if (this.IsExpired())
+            if (IsExpired())
             {
-                throw new UnauthorizedAccessException($"The specified SAS token is expired on {this.ExpiresOn}.");
+                throw new UnauthorizedAccessException($"The specified SAS token is expired on { ExpiresOn}.");
```
Comment (barustum) [inline iothub/device/src/Common/Security/SharedAccessSignature.cs]: space after ExpiresOn

File: iothub/device/src/Common/Exceptions/IotHubException.cs (hunk 4 above)
Comment (drwill-ms) [reply iothub/device/src/Common/Exceptions/IotHubException.cs]: no, good catch

File: iothub/device/src/Common/Security/SharedAccessSignature.cs [hunk 6]
```diff
@@ -12,105 +12,69 @@ namespace Microsoft.Azure.Devices.Client
...
         }
 
-        public string IotHubName
-        {
-            get
-            {
-                return this.iotHubName;
-            }
-        }
+        public string IotHubName { get; private set; }
```
Comment (drwill-ms) [inline iothub/device/src/Common/Security/SharedAccessSignature.cs]: readonly

File: provisioning/transport/amqp/src/ProvisioningTransportHandlerAmqp.cs [hunk 7]
```diff
@@ -191,66 +196,88 @@ private static async Task CreateLinksAsync(AmqpClientConnection connection, stri
...
 
         private static DeviceRegistrationResult ConvertToProvisioningRegistrationResult(
             Models.DeviceRegistrationResult result)
         {
-            var status = ProvisioningRegistrationStatusType.Failed;
-            Enum.TryParse(result.Status, true, out status);
-
-            var substatus = ProvisioningRegistrationSubstatusType.InitialAssignment;
-            Enum.TryParse(result.Substatus, true, out substatus);
+            Enum.TryParse(result.Status, true, out ProvisioningRegistrationStatusType status);
```
Comment (vinagesh) [inline provisioning/transport/amqp/src/ProvisioningTransportHandlerAmqp.cs]: Default is Unassigned so will this change the behavior now?

Comment (drwill-ms) [reply provisioning/transport/amqp/src/ProvisioningTransportHandlerAmqp.cs]: The .net contract for `out` parameters is that they must be assigned to. So our previous assignment of Failed didn't matter.

Does that answer your question?

**Pull request merged**